client/src/controllers/RevealController.ts

```diff
@@ -0,0 +1,311 @@
+ * </section>
+ */
+export class RevealController extends Controller<HTMLElement> {
+  static classes = [
+    'closed',
+    'closeIcon',
+    'initial',
+    'opened',
+    'openedContent',
+    'openIcon',
+  ];
+
+  static targets = ['content', 'toggle'];
+
+  static values = {
+    closed: { default: false, type: Boolean },
+    peeking: { default: false, type: Boolean },
+    peekTarget: { default: '', type: String },
+  };
+
+  declare closedValue: boolean;
+  declare peekingValue: boolean;
+
+  declare readonly closedClass: string;
+  declare readonly closedClasses: string[];
+  declare readonly closeIconClass: string;
+  declare readonly contentTarget: HTMLElement;
+  declare readonly contentTargets: HTMLElement[];
+  declare readonly hasClosedClass: boolean;
+  declare readonly hasCloseIconClass: string;
+  declare readonly hasContentTarget: boolean;
+  declare readonly hasOpenIconClass: string;
+  declare readonly hasToggleTarget: boolean;
+  declare readonly initialClasses: string[];
+  declare readonly openedClasses: string[];
+  declare readonly openedContentClasses: string[];
+  declare readonly openIconClass: string;
+  declare readonly peekTargetValue: string;
+  declare readonly toggleTarget: HTMLButtonElement;
+  declare readonly toggleTargets: HTMLButtonElement[];
+
+  cleanUpPeekListener?: () => void;
+
+  /**
+   * Ensure that targets are correctly set up, as this runs
+   * before the initial call to `closedValueChanged`.
+   * If there is no content element set, try to set one
+   * via `aria-controls` on the toggle, this also ensures
+   * that the toggleTarget has been set before the controller
+   * connects.
+   */
+  initialize() {
+    const toggleTarget = this.toggleTarget;
+
+    if (!this.hasContentTarget) {
+      const ariaControlsSelectors = (
+        toggleTarget.getAttribute('aria-controls') || ''
+      )
+        .split(' ')
+        .filter(Boolean)
+        .map((id) => `#${id}`)
+        .join(',');
+
+      const contentElements = ariaControlsSelectors
+        ? document.querySelectorAll(ariaControlsSelectors)
+        : [];
+
+      if (!contentElements.length) {
+        const error = `data-${this.identifier}-target="content" must be set or aria-controls must target a valid element.`;
+        throw new Error(error);
+      }
+
+      contentElements.forEach((content) => {
+        content.setAttribute(`data-${this.identifier}-target`, 'content');
+      });
+    }
+
+    // Only attempt to set to closed if not already closed via initial value
+    const isClosed = this.closedValue;
+    if (isClosed) return;
+
+    const ariaExpanded = toggleTarget.getAttribute('aria-expanded');
+    const isClassNameClosed = this.hasClosedClass
+      ? this.element.classList.contains(this.closedClass)
+      : false;
+
+    if (isClassNameClosed && (!ariaExpanded || ariaExpanded === 'false')) {
+      this.closedValue = true;
+    } else if (ariaExpanded === 'false') {
+      this.closedValue = true;
+    }
```

Comment:
This is mostly to set up for the panels use case where you can add the `collapse` class and it will load in the DOM as collapsed.

https://github.com/wagtail/wagtail/blob/main/client/src/includes/panels.ts#L67

It's confusing I know but it will be required for the Panels usage alongside dynamic usage in StreamFields.

I can remove this for now and we can just discuss the implementation when ready for the Panels work.

Reply:
This has been removed.

Reply:
Ahh yes, sorry, I completely forgot about the fact that we support the `collapse` class for panels. But yep, while it's necessary for panels, we can focus on that later.

Maybe we can look into alternatives such as providing the appropriate Stimulus attribute from the Python side if the `collapse` class is used, or a different approach entirely e.g. a `collapsed` flag on the `Panel` instance itself, with the Stimulus attributes handled in the base class. In any case, it's easier to focus on exploring those approaches later than trying to fit it in here. Thanks for the update!